File: binary_search/python/binary_search.py
def binary_search(lista:list, valor) -> int:
    inicio = 0
    fim = len(lista) - 1  # Pegando o indice da ultima posição da lista.
    
    # Enquanto o incio for menor que o final da lista segue a interação.
    while inicio <= fim:
        meio = (inicio + fim) // 2 # Pegando o indice do meio da lista.
        palpite = lista[meio] # Pegando o valor do indice no meio da lista.

        # Compara se o valor pego é igual ao procurado, caso verdadeiro retorna o indice 
        if palpite == valor:
            return meio
        
        # Verifica se o valor é maior que o procurado, caso verdadeiro o final da lista é reduzido.
        if palpite > valor:
            fim = meio - 1

        # Verifica se o valor é menor que o procurado, caso verdadeiro o inicio da lista é aumentado.
        if palpite < valor:
            inicio = meio + 1

    return - 1

File: binary_search/c/binary_search.c
#include <stdio.h>

int binary_search(int list[], int size, int value){
    int inicio, fim, meio, chute;

    inicio = 0;
    fim = size - 1;

    while (inicio <= fim) {
        meio = (inicio + fim);
        chute = list[meio];

        if (chute == value) {
            return meio;
        }

        if (chute > value) {
            fim = meio -1;
        }

        if (chute < value) {
            inicio = meio +1;
        }
    }   
    return -1;
}

/* Função para printar o Array */
void print_array(int list[], int size){
    for (int i=0; i<size; i++){
        printf("%d ", list[i]);
    }
    printf("\n");
}


int main(void){

    int date[] = {2, 3, 4, 5, 8, 9, 11, 13, 15, 16, 17, 19, 20, 24, 24};
    int size =  sizeof(date) / sizeof(date[0]);

    print_array(date, size);

    int index = binary_search(date, size, 5);

    printf("\nO indice do valor é %d\n", index);

    return 0;
}

